### DISCUSSION TITLE
MsgSpecJSONResponse: JSONResponse support using msgspec

### DISCUSSION THREAD

**Discussion by rapidrabbit76** [fastapi/fastapi#10616] [Show and tell]
Hello, I mainly create machine learning APIs using FastAPI.
As I often have to create large amounts of JSON responses, I came across msgspec while refactoring to improve the response speed of my personal service.

There is no similar content in issues, discussions, and PR, so I would like to introduce it here first.

The existing project has responses in ORJSON and UJSON, so they are created and used in the same format.

```python
# fastapi/responses.py
...
try:
    import msgspec
except ImportError:  # pragma: nocover
    msgspec = None  # type: ignore

...

class MsgSpecJSONResponse(JSONResponse):
    """
    JSON response using the high-performance msgspec library to serialize data to JSON.
    """

    def render(self, content: Any) -> bytes:
        assert msgspec is not None, "msgspec must be installed to use MsgSpecJSONResponse"
        return msgspec.json.encode(content)

```

For use, enter the "default_response_class" of the API the same as ORJSON and UJSON.

```python
from fastapi import APIRouter
router = APIRouter(default_response_class=MsgSpecJSONResponse)
```

**Comment by tiangolo** [comment]
The solution is precisely what you did, create a custom class with your serialization and use it. 🤓 ☕ 

```Python
from typing import Any

from fastapi import APIRouter


class MsgSpecJSONResponse(JSONResponse):
    def render(self, content: Any) -> bytes:
        return msgspec.json.encode(content)


router = APIRouter(default_response_class=MsgSpecJSONResponse)
```

Creating a custom class is just 3 lines, and allows you all the flexibility you could need to use any library you want, so I prefer to teach how to do that: https://fastapi.tiangolo.com/advanced/custom-response/#custom-response-class than to include each possible response in the library.

**Reply by ShafSpecs** [reply]
How would this work with a `WebSocket`? Looking at `msgspec`, it beats what `fastapi` currently offers for json responses by far in terms of benchmarks (under a ms). Do I pass in `default_response_class=` or `send_text` and pass in my encoded json or is there another way?

And is FastAPI looking to use msgspec for some of its more generic encodings/decodings?

**Comment by yqiang** [comment]
Since `msgspec` has a `msgspec.Struct` type that is somewhat similar to `BaseModel` from pydantic, is it possible to return the model directly instead of converting it to a dict first?

**Reply by dorinclisu** [reply]
Look at Litestar for that, msgspec is a first class citizen. 
https://github.com/litestar-org/litestar
